core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -144,288 +141,181 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
       return invokeNext(ctx, command);
    }
 
-   protected final CompletableFuture<InternalCacheEntry> retrieveFromProperSource(Object key, InvocationContext ctx,
-         FlagAffectedCommand command, boolean isWrite) throws Exception {
-      return doRetrieveFromProperSource(key, null, -1, ctx, command, isWrite);
-   }
-
-   private CompletableFuture<InternalCacheEntry> doRetrieveFromProperSource(Object key, InternalCacheEntry value,
-         int lastTopologyId, InvocationContext ctx, FlagAffectedCommand command, boolean isWrite) {
-      if (value != null)
-         return CompletableFuture.completedFuture(value);
-
-      final CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-      final int currentTopologyId = cacheTopology.getTopologyId();
+   protected final CompletableFuture<Void> remoteGet(InvocationContext ctx, AbstractDataCommand command,
+                                                     boolean isWrite) throws Exception {
+      final CacheTopology cacheTopology = checkTopologyId(command);
```

Comment:
Replying with an `OutdatedTopologyResponse` instead of the exception sounds like a good thing, if only to minimize the overhead of all those stack traces. But I still don't see a good reason for the reads to block at all... 

Say the sender has topology TS and the receiver has topology TR. 

If TS < TR and the receiver is a read owner in TR, the receiver knows it has the proper value, so it can return it. If it's only a write owner or not an owner at all, it can return an OTR. But if the sender receives an OTR and it still has topology TS, it knows who to ask for the value immediately: the write owners in TS. The coordinator can't finish a rebalance without installing the rebalance-in-progress topology on all the members, so the sender must install a newer rebalance-in-progress topology before the receiver can see a topology where the write owners in TS to all become non-owners.

If TS > TR, the receiver can only be a read owner for more keys in topology TS if it was a write owner of those keys in topology TR and it had already sent the REBALANCE_CONFIRM(TR) command to the coordinator before the sender installed topology TR. But that means the receiver already has the requested key, so blocking in STI to receive topology TS delays the read but doesn't make it any safer.

For non-staggered gets, when the response filter is called with an OTR it could compute what the new recipients would be (either from the OTR or from the sender's current topology) and decide whether to stop the current request based on how much of an overlap there is. But for staggered gets I think the penalty for ending the request early and starting a new one is minimal, so I'd always cancel the request when receiving the first OTR.